### PR TITLE
Free widgets

### PR DESCRIPTION
**Pull request by markrwilliams**
Widgets don't work with free functions that don't have a first positional argument, or with instance methods.

A first positional argument is required because the returned `wrapper` has `instance` as its first argument:

https://github.com/twisted/klein/blob/67580644a53201332ac48a0fa18769d19fad6338/src/klein/_plating.py#L167

Instance methods don't work because the `wrapper` can't forward the bound `self` to its invocation of `function`:

https://github.com/twisted/klein/blob/67580644a53201332ac48a0fa18769d19fad6338/src/klein/_plating.py#L168

because it's assigned as an attribute on the function:

https://github.com/twisted/klein/blob/67580644a53201332ac48a0fa18769d19fad6338/src/klein/_plating.py#L170

This PR introduces a descriptor that addresses both by binding the function it wraps on attribute access.

### CODE REVIEW

**Comment by glyph** [comment]
test results don't look…great…

File: src/klein/_plating.py [hunk 1]
```diff
@@ -161,11 +163,45 @@ def _elementify(self, instance, to_fill_with):
                              boundInstance=instance,
                              presentationSlots=self._presentationSlots)
 
+    @attr.s
+    class _Widget(object):
+        """
+        Implementation of L{Plating.widgeted}.  This is a L{callable}
```
Comment (wsanchez) [inline src/klein/_plating.py]: Callable, not callable, right?

File: src/klein/_plating.py [hunk 2]
```diff
@@ -161,11 +163,45 @@ def _elementify(self, instance, to_fill_with):
...
+        L{PlatedElement}.
+        """
+        _plating = attr.ib()
+        _function = attr.ib()
+        _instance = attr.ib()
+
+        def __call__(self, *args, **kwargs):
+            return self._function(*args, **kwargs)
+
+        def __get__(self, obj, type_=None):
```
Comment (wsanchez) [inline src/klein/_plating.py]: Can we obj, type ➜ instance, owner to match __get__signature in the python docs?

File: src/klein/_plating.py [hunk 3]
```diff
@@ -161,11 +163,45 @@ def _elementify(self, instance, to_fill_with):
...
+            )
+
+        def widget(self, *args, **kwargs):
+            """
+            Construct a L{PlatedElement} the rendering of this widget.
+            """
+            data = self._function(*args, **kwargs)
+            return self._plating._elementify(self._instance, data)
+
+        def __getattr__(self, attr):
```
Comment (wsanchez) [inline src/klein/_plating.py]: attr ➜ name to match documented signature

File: src/klein/_plating.py (hunk 1 above)
Comment (markrwilliams) [reply src/klein/_plating.py]: Good point - it needs to implement `__get__`.

**Pull request merged**